packages/zoe/src/types.js

```diff
@@ -28,103 +25,104 @@
- * offer) as well as payments to be escrowed by Zoe.  If either the proposal or payments
- * would be empty, indicate this by omitting that argument or passing undefined, rather
- * than passing an empty record.
- *
- * The proposal has three parts: `want` and `give` are used
- * by Zoe to enforce offer safety, and `exit` is used to specify
- * the particular payout-liveness policy that Zoe can guarantee.
- * `want` and `give` are objects with keywords as keys and amounts
- * as values. `paymentKeywordRecord` is a record with keywords as keys,
- * and the values are the actual payments to be escrowed. A payment
- * is expected for every rule under `give`.
- *
- * @property {(offerHandle: OfferHandle) => boolean} isOfferActive
- * @property {(offerHandles: OfferHandle[]) => OfferRecord[]} getOffers
- * @property {(offerHandle: OfferHandle) => OfferRecord} getOffer
- * @property {(offerHandle: OfferHandle) => Notifier<Allocation|undefined>} getOfferNotifier
- * Get a notifier (see `@agoric/notify`) for the offer's reallocations.
- * @property {(offerHandle: OfferHandle, brandKeywordRecord?: BrandKeywordRecords) => Allocation} getCurrentAllocation
- * @property {(offerHandles: OfferHandle[], brandKeywordRecord[]?: BrandKeywordRecords) => Allocation[]} getCurrentAllocations
- * @property {(installationHandle: InstallationHandle) => SourceBundle} getInstallation
- * Get the source code for the installed contract. Throws an error if the
- * installationHandle is not found.
- *
- * @typedef {Object} CompleteObj
- * @property {() => void} complete attempt to exit the contract and return a refund
+ * registering it with Zoe. Returns an installation.
  */
 
 /**
- * @template OC - the offer outcome
- * @typedef {Object} OfferResultRecord This is returned by a call to `offer` on Zoe.
- * @property {Promise<OfferHandle>} offerHandle
- * @property {Promise<PaymentPKeywordRecord>} payout A promise that resolves
- * to a record which has keywords as keys and promises for payments
- * as values. Note that while the payout promise resolves when an offer
- * is completed, the promise for each payment resolves after the remote
- * issuer successfully withdraws the payment.
- *
- * @property {Promise<OC>} outcome Note that if the offerHook throws,
- * this outcome Promise will reject, but the rest of the OfferResultRecord is
- * still meaningful.
- *
- * @property {CompleteObj} [completeObj]
- * completeObj will only be present if exitKind was 'onDemand'
+ * @typedef {Object} UserSeat
+ * @property {() => Promise<Allocation>} getAllocation
+ * @property {() => Promise<ProposalRecord>} getProposal
+ * @property {() => Promise<PaymentPKeywordRecord>} getPayouts
+ * @property {(keyword: Keyword) => Promise<Payment>} getPayout
+ * @property {() => Promise<OfferResult>} getOfferResult
+ * @property {() => void} exit
+ */
+
+/**
+ * @callback Offer
+ * @param {Invitation|Promise<Invitation>} invitation
```

Comment:
Use @michaelfig 's type `PromiseOrNot` to replace `{Invitation|Promise<Invitation>}` with `{PromiseOrNot<Invitation>}`. Likewise for other `{X|Promise<X>}` or `{X|PromiseLike<X>}` occurrences.

`PromiseOrNot` is currently at https://github.com/Agoric/agoric-sdk/blob/1d2925ad640cce7b419751027b44737bd46a6d59/packages/notifier/src/types.js#L1-L4 but should be moved somewhere appropriate for wide reuse. @michaelfig , where?

Reply:
@michaelfig the reason `PromiseOrNot<T>` comes up a lot is that it is the general form of a "reference to T", where a promise is a reference to a T that may be separated in time or space. In E, a direct reference is a "near reference". These others are "eventual references". See http://www.erights.org/elib/concurrency/refmech.html

Or a "pointer to T", or "something that designates a T". Can we come up with a name describing the purpose, rather than one enumerating the choices? Is `Ref<T>` already taken or have a different meaning?

Reply:
Ok, I will need to rebase in order to have access

Reply:
See #1383 . @katelynsills I recommend ignoring this one until #1383 lands.